### PR TITLE
cilium: allow encryption/decryption to coexist with bpf_host logic

### PR DESCRIPTION
**Pull request by jrfastab**
We have added logic in the bpf_host that is attached to egress/ingress network facing device. This is the same device that is used by bpf_network for encryption. The result is bpf_network is loaded by bpf_init.sh and then bpf_host based programs are loaded by golang bpf loader replacing the bpf_network program. This results in traffic drops because encryption rules exist in xfrm datapath, but BPF side is not correctly implementing them. To fix make bpf_host do correct decrypt logic.

Additionally, workaround a kernel regression by simplifing the encryption/decryption rules in the encrypt-node case. Kernels 4.19 and earlier appear to work correctly with the current setup, but newer kernels (at least 5.4) is broken with the current setup. The new cleaner approach appears to work on all kernels testsed

### CODE REVIEW

File: pkg/datapath/linux/node.go
Comment (christarazi) [inline pkg/datapath/linux/node.go]: Am I reading this correctly? We check if the node has IPSec, and then check if it **doesn't** have IPSec? Maybe this was a typo?

**Review by christarazi — COMMENTED** [review]
Looks good overall, potential typo down below.

File: pkg/datapath/linux/node.go
Comment (jrfastab) [reply pkg/datapath/linux/node.go]: its a typo ;) thanks. I forgot to commit the fix.

Comment (jrfastab) [reply pkg/datapath/linux/node.go]:
```
if n.nodeConfig.EnableIPSec && !n.subnetEncryption() && !n.nodeConfig.EncryptNode {
```
is the intended line.

**Review by christarazi — APPROVED** [review]
LGTM. PR title has `coexit`, which I think you meant `coexist` :)

File: bpf/lib/encrypt.h
Comment (pchaigno) [inline bpf/lib/encrypt.h:58]: Can we keep the comment on this statement? It's a bit hard to understand at first without context.

File: pkg/datapath/linux/node.go
Comment (pchaigno) [inline pkg/datapath/linux/node.go]: Should this be in its own PR to be backported to 1.8 and 1.7?

File: test/k8sT/DatapathConfiguration.go
Comment (pchaigno) [inline test/k8sT/DatapathConfiguration.go:573]: I think you will need to revert https://github.com/cilium/cilium/commit/8cbd3096a829940035eea95e1e6b10980b07e46e for this test to pass.

**Review by pchaigno — CHANGES_REQUESTED** [review]
LGTM overall! A couple of comments below, one of them I think you will need to address for the e2e tests to pass.

I'm currently testing if IPsec now works with the host firewall and the bandwidth manager. I'll have the answer tomorrow.

File: test/k8sT/DatapathConfiguration.go
Comment (jrfastab) [reply test/k8sT/DatapathConfiguration.go:573]: Yes agree. I must have had a tree without that patch though.Interesting.

File: bpf/lib/encrypt.h
Comment (jrfastab) [reply bpf/lib/encrypt.h:58]: sure I'll add it back.

Comment (jrfastab) [reply bpf/lib/encrypt.h:58]: Actually, what comment are you referring to here? Seems the "Decrypt key is determined by SPI" is still there.

File: pkg/datapath/linux/node.go
Comment (jrfastab) [reply pkg/datapath/linux/node.go]: Makes sense we can pull it out.

**Comment by jrfastab** [comment]
First patch pulled into this PR https://github.com/cilium/cilium/pull/13241 once we get that merged I'll pop the first patch off this series.

File: bpf/lib/encrypt.h
Comment (pchaigno) [reply bpf/lib/encrypt.h:58]: I'm referring to:
> ```
> /* We are going to pass this up the stack for IPsec decryption
>  * but eth_type_trans may already have labeled this as an
>  * OTHERHOST type packet. To avoid being dropped by IP stack
>  * before IPSec can be processed mark as a HOST packet.
>  */
> ```

**Comment by pchaigno** [comment]
retest-4.9

**Review by aanm — APPROVED** [review]
LGTM for the agent parts

File: bpf/lib/encrypt.h
Comment (jrfastab) [reply bpf/lib/encrypt.h:58]: OK will add this back.

Comment (jrfastab) [reply bpf/lib/encrypt.h:58]: done.

**Comment by jrfastab** [comment]
@qmonnet should we remove this from the checkpatch tests as well?
```
WARNING:FILE_PATH_CHANGES: added, moved or deleted file(s), does MAINTAINERS need updating?
#156: 
new file mode 100644
```

**Comment by jrfastab** [comment]
Also for now I've kept ipsec blocked by bandwidth manager, because I haven't walked through the details yet to see if it is also working. But, it should be getting closer now.

**Comment by qmonnet** [comment]
> @qmonnet should we remove this from the checkpatch tests as well?

Agreed, it's on the todo list (see [related issue](https://github.com/cilium/cilium/issues/12467)). I'll try to free some cycles to work on this soon.

**Review by pchaigno — APPROVED** [review]
> Also for now I've kept ipsec blocked by bandwidth manager, because I haven't walked through the details yet to see if it is also working. But, it should be getting closer now.

It's not. I tested this with both the bandwidth manager and the host firewall. In the bandwidth manager's case, the rate limiting wasn't enforced. In the host firewall's case, IIRC the cluster didn't become healthy even without host policies.

I'll try to take a look at those again after the code freeze.

**Comment by pchaigno** [comment]
test-me-please

**Comment by jrfastab** [comment]
Consulted data around failure the failure in kernel-netnext seems to be happening in multiple places, perhaps a flake?

Suite-k8s-1.12.K8sPolicyTest Multi-node policy test validates ingress CIDR-dependent L4 With host policy Connectivity is restored after importing ingress policy

**Comment by jrfastab** [comment]
@pchaigno I assume for bandwidth manager you had to disable the check to not allow it with encryption. But, host firewall do we need a check?

**Comment by pchaigno** [comment]
> Consulted data around failure the failure in kernel-netnext seems to be happening in multiple places, perhaps a flake?
> 
> Suite-k8s-1.12.K8sPolicyTest Multi-node policy test validates ingress CIDR-dependent L4 With host policy Connectivity is restored after importing ingress policy

I suspect this might be the same underlying bug as is affecting #12596. The two tests are very similar, just the above tests a host policy instead of "pod policy".

> @pchaigno I assume for bandwidth manager you had to disable the check to not allow it with encryption. But, host firewall do we need a check?

Yes, we have check for each of these when combined with IPSec. I removed them to be able to test.

**Comment by jrfastab** [comment]
So unless I'm mistaken we have a checkpatch warning thats not needed, `WARNING:FILE_PATH_CHANGES: added, moved or deleted file(s), does MAINTAINERS need updating?` and a flake? So we should be OK to merge.

**Comment by borkmann** [comment]
Would it now be possible to merge bpf_network.c logic into bpf_host.c and remove the former entirely?

**Comment by jrfastab** [comment]
@borkmann I believe so thats the next step, but first need to fix above error

**Pull request merged**
